### ISSUE TITLE
Use 'mail' attribute to login instead of 'userprincipalname' does not works.

### ISSUE THREAD

**Issue by paolobasso99** [Adldap2/Adldap2-Laravel#330]
- Laravel Version: 5.4.*
- Adldap2-Laravel Version: ^3.0
- PHP Version: 7.0

Hi, if I configure `adldap_auth.php` as:

```
'usernames' => [

        'ldap' => 'userprincipalname',

        'eloquent' => 'email',

    ],
```
everything works fine.
If I try to use `mail` instead of `userprincipalname` the credentials doesn't match. How can I fix it?

**Comment by stevebauman** [comment]
Hi @paolobasso99, this sounds like an issue with your AD server. The user needs to contain a `mail` attribute on their user account.

Are you using ActiveDirectory?

**Comment by paolobasso99** [comment]
Yes, if i do `Adldap::search()->users()->find('Name Surname')->mail;` it return the mail.
Also if I edit the `DatabaseUserProvider` adding a die and dump i get the User retrived using 'mail'.
```
public function retrieveByCredentials(array $credentials)
    {
        // Retrieve the LDAP user who is authenticating.
        $user = $this->getResolver()->byCredentials($credentials);
        dd($user);
        if ($user instanceof User) {
            // Set the currently authenticating LDAP user.
            $this->user = $user;

            $this->handleDiscoveredWithCredentials($user);

            // Import / locate the local user account.
            return $this->getImporter()
                ->run($user, $this->createModel(), $credentials);
        }

        if ($this->isFallingBack()) {
            return $this->fallback->retrieveByCredentials($credentials);
        }
    }
```

**Comment by stevebauman** [comment]
Ok, can you debug further and see where it fails to login your user account?

**Comment by 36864** [comment]
I'm starting a new project in which I need to use LDAP for the first time. I ran into trouble trying to authenticate the user by `mail` rather than a full DN using the `HasLdapUser` trait on my `User` model as described in https://github.com/Adldap2/Adldap2-Laravel/blob/master/docs/auth/binding.md
I've managed to work around my issue for now. It's not a great solution, but it'll work until I can come back and make it better.
As far as I can tell, when trying to authenticate a user, the method that is called is `validateCretendials`, which calls `authenticate`, which simply attempts a `bind()` on your LDAP using what you define as a `username` and the `password` provided by your user.
In my case, users have a `uid` key and I want them to log in using their e-mail address. Adding a step to the `authenticate` method  in `Adldap\Laravel\Auth\Resolver` solved my issue. 

```php
public function authenticate(User $user, array $credentials = [])
{
    $username = $user->getFirstAttribute($this->getLdapUsername());
    $username = Adldap::search()->select('uid')->findBy('mail', $username)->uid[0];          
    return $this->provider->auth()->attempt($username, $credentials['password']);
}
```
Here's my understanding of the problem, which I believe to be similar to yours.
The ldap username config option only serves to change what the library will use as the username field when attempting to bind to your LDAP server. It does not tell it which attribute to check and as far as I can tell, you cannot bind through an arbitrary attribute, such as `mail`.
In order to authenticate via the `mail` attribute, you'll have to first find the user to whom the email belongs and then attempt a bind using that user's dn/uid.

This is working on OpenLDAP.

**Comment by diegochaves** [comment]
Same problem as @36864 and I fixed the same way.
My debug was based on tcpdumping between my PC and ActiveDirectory Server.
The problems resides on bindRequest using mail attribute.

**Comment by stevebauman** [comment]
Thanks guys, this is actually being resolved in `v4.0`, with the ability to configure a `discover` and `authenticate` attribute:

https://github.com/Adldap2/Adldap2-Laravel/blob/master/src/Config/auth.php#L95-L127

**Comment by stevebauman** [comment]
Released in `v4.0`.
